scripts/download_hash.sh

```diff
@@ -187,73 +256,85 @@ function _get_checksum {
-["skopeo_binary"]="$(printf "$github_releases_url" "lework/skopeo-binary" "skopeo-$os-$arch")"
+["containerd_archive"]="$(printf "$github_releases_url" "containerd/containerd" "containerd-${version#v}-$os-$arch.tar.gz.sha256sum")"
+["skopeo_binary"]="$(printf "$github_releases_url" "lework/skopeo-binary" "skopeo-$os-$arch.sha256")"
 ["yq"]="$(printf "$github_releases_url" "mikefarah/yq" "yq_${os}_$arch")"
 )
 
     mkdir -p "$(dirname $target)"
     [ -f "$target" ] || curl -LfSs -o "${target}" "${urls[$binary]}"
-    sha256sum ${target} | awk '{print $1}'
-}
-
-function main {
-    mkdir -p "$(dirname "$checksums_file")"
-    echo "---" | tee "$checksums_file"
-    get_checksums crictl $(get_versions github_tags kubernetes-sigs/cri-tools 4)
-    get_checksums crio_archive $(get_versions github_tags cri-o/cri-o)
-    kubernetes_versions=$(get_versions github_tags kubernetes/kubernetes 25)
-    echo "# Checksum" | tee --append "$checksums_file"
-    echo "# Kubernetes versions above Kubespray's current target version are untested and should be used with caution." | tee --append "$checksums_file"
-    get_k8s_checksums kubelet $kubernetes_versions
-    get_checksums kubectl $kubernetes_versions
-    get_k8s_checksums kubeadm $kubernetes_versions
-    get_checksums etcd_binary $(get_versions github_tags etcd-io/etcd)
-    get_checksums cni_binary $(get_versions github_tags containernetworking/plugins)
-    calico_versions=$(get_versions github_tags projectcalico/calico 20)
-    get_checksums calicoctl_binary $calico_versions
-    get_checksums ciliumcli_binary $(get_versions github_tags cilium/cilium-cli 10)
-    get_calico_crds_archive_checksums $calico_versions
-    get_krew_archive_checksums $(get_versions github_tags kubernetes-sigs/krew 2)
-    get_checksums helm_archive $(get_versions github_tags helm/helm)
-    get_checksums cri_dockerd_archive $(get_versions github_tags Mirantis/cri-dockerd)
-    get_checksums runc $(get_versions github_tags opencontainers/runc 5)
-    get_checksums crun $(get_versions github_tags containers/crun)
-    get_checksums youki $(get_versions github_tags containers/youki)
-    get_checksums kata_containers_binary $(get_versions github_tags kata-containers/kata-containers 10)
-    gvisor_versions=$(get_versions github_tags google/gvisor 0 "import sys,re;tags=[tag[8:16] for tag in sys.stdin if re.match('^release-?(0|[1-9]\d*)\.(0|[1-9]\d*)$',tag)];print(' '.join(tags[:9]))")
-    get_checksums gvisor_runsc_binary $gvisor_versions
-    get_checksums gvisor_containerd_shim_binary $gvisor_versions
-    get_checksums nerdctl_archive $(get_versions github_tags containerd/nerdctl)
-    get_containerd_archive_checksums $(get_versions github_tags containerd/containerd 30)
-    get_checksums skopeo_binary $(get_versions github_tags lework/skopeo-binary)
-    get_checksums yq $(get_versions github_tags mikefarah/yq)
+    if echo "${urls[$binary]}" | grep -qi sha256sum; then
```

Comment:
HI @MrFreezeex 

Based on my original understanding, the goal of computing the hash is to avoid security risks like middle-man attacks. 

For example, a firewall between the download machine and github.com changes the download file. The kubespray can compare the sha256sum between the downloaded file and code and avoid security risks. The risks cannot be avoided if the kubespray downloads the sha256sum file instead of computing it.

Thanks :-) 🤝🤝🤝

Reply:
Hi!

> For example, a firewall between the download machine and github.com changes the download file. The kubespray can compare the sha256sum between the downloaded file and code and avoid security risks. The risks cannot be avoided if the kubespray downloads the sha256sum file instead of computing it.

I don't see how it is any different. If the file you download with this script is compromised whether it's a hash, a hashsum or the file it will put a bad hash in the kubespray checksum file. I don't believe there is any difference in term of security but it does have some performance benefits for sure.

Reply:
@yankay the goal would be to record the hash in kubespray so that users can check the hash when getting binaries. But when we get the hash, we're getting them from the same source than the binaries, so if one is compromised, so is the other.

If we want stronger guarantees about the file provenance we could instead look at [signed Kubernetes artifacts](https://kubernetes.io/docs/tasks/administer-cluster/verify-signed-artifacts/) but that's another subject